tests/test_grpc_server.py

```diff
@@ -71,6 +73,271 @@ def test_read_object(self):
+        self.assertEqual(blob.bucket, "projects/_/buckets/bucket-name")
+
+    def test_object_write_invalid_request(self):
+        context = unittest.mock.Mock()
+        context.abort = unittest.mock.MagicMock()
+        write = self.grpc.WriteObject([], context)
+        context.abort.assert_called_once()
+        self.assertIsNone(write)
+
+    def test_object_write_incomplete(self):
+        QUANTUM = 256 * 1024
+        media = TestGrpc._create_block(2 * QUANTUM + QUANTUM / 2).encode("utf-8")
+
+        offset = 0
+        content = media[0:QUANTUM]
+        r1 = storage_pb2.WriteObjectRequest(
+            write_object_spec=storage_pb2.WriteObjectSpec(
+                resource={
+                    "name": "object-name",
+                    "bucket": "projects/_/buckets/bucket-name",
+                },
+            ),
+            write_offset=offset,
+            checksummed_data=storage_pb2.ChecksummedData(
+                content=content, crc32c=crc32c.crc32c(content)
+            ),
+            finish_write=False,
+        )
+
+        context = unittest.mock.Mock()
+        context.abort = unittest.mock.MagicMock()
+        write = self.grpc.WriteObject([r1], context)
+        context.abort.assert_called_once()
+        self.assertIsNone(write)
+
+    def test_resumable_write(self):
+        start = self.grpc.StartResumableWrite(
+            storage_pb2.StartResumableWriteRequest(
+                write_object_spec=storage_pb2.WriteObjectSpec(
+                    resource=storage_pb2.Object(
+                        name="object-name", bucket="projects/_/buckets/bucket-name"
+                    )
+                )
+            ),
+            context="fake-context",
+        )
+        self.assertIsNotNone(start.upload_id)
+
+        def streamer():
+            media = TestGrpc._create_block(517 * 1024).encode("utf-8")
+            step = 256 * 1024
+            end = min(step, len(media))
+            content = media[0:end]
+            # The first message is special, it should have a an upload id.
+            yield storage_pb2.WriteObjectRequest(
+                upload_id=start.upload_id,
+                write_offset=0,
+                checksummed_data=storage_pb2.ChecksummedData(
+                    content=content, crc32c=crc32c.crc32c(content)
+                ),
+                finish_write=(end == len(media)),
+            )
+
+            for offset in range(step, len(media), step):
+                end = min(offset + step, len(media))
+                content = media[offset:end]
+                yield storage_pb2.WriteObjectRequest(
+                    write_offset=offset,
+                    checksummed_data=storage_pb2.ChecksummedData(
+                        content=content, crc32c=crc32c.crc32c(content)
+                    ),
+                    finish_write=(end == len(media)),
+                )
+
+        write = self.grpc.WriteObject(streamer(), "fake-context")
```

Comment:
I haven't user generators much, but I think we need a for loop in order to generate (and use) each `WriteObjectRequest`

IIUC, `streamer()` is just a generator object. To yield each `WriteObjectRequest`, we either need to utilize a for loop or call `next()` on the generator object.

Reply:
Oh wait does `WriteObject` consume a stream-like generator?

Reply:
So it works right?  I just tested with:

```py
def foo():
...     yield 1
...     yield 2
...     for i in range(10, 15): yield i
... 
>>> 
>>> 
>>> g = foo()
>>> l = list(g)
>>> print(l)
[1, 2, 10, 11, 12, 13, 14]
```

That is what I expected.

Reply:
Yep looks good! Sorry about the confusion 😂

Reply:
Thanks to you.